restli-tools/src/main/java/com/linkedin/restli/tools/snapshot/check/PegasusSchemaSnapshotCompatibilityChecker.java

```diff
@@ -0,0 +1,297 @@
+
+    File preSchemaFile = new File(prevPegasusSchemaPath);
+    File currSchemaFile = new File(currentPegasusSchemaPath);
+
+    DataSchema preSchema = null;
+    try
+    {
+      preSchema = parseSchema(preSchemaFile);
+    }
+    catch(FileNotFoundException e)
+    {
+      constructCompatibilityMessage(CompatibilityMessage.Impact.OLD_READER_IGNORES_DATA,
+          "New schema %s is created.", currentPegasusSchemaPath);
+    }
+
+    DataSchema currSchema = null;
+    try
+    {
+      currSchema = parseSchema(currSchemaFile);
+    }
+    catch(FileNotFoundException e)
+    {
+      constructCompatibilityMessage(CompatibilityMessage.Impact.BREAKS_OLD_READER,
+          "Schema %s is removed.", prevPegasusSchemaPath);
+    }
+
+    if (preSchema == null || currSchema == null)
+    {
+      return;
+    }
+
+    CompatibilityOptions compatibilityOptions = new CompatibilityOptions().setMode(compatMode).setAllowPromotions(true);
+    CompatibilityResult result = CompatibilityChecker.checkCompatibility(preSchema, currSchema, compatibilityOptions);
+
+    if (!result.getMessages().isEmpty())
+    {
+      result.getMessages().forEach(message -> _infoMap.addModelInfo(message));
+    }
+  }
+
+  private void constructCompatibilityMessage(CompatibilityMessage.Impact impact, String format, Object... args)
+  {
+    CompatibilityMessage message = new CompatibilityMessage(new Object[]{}, impact, format, args);
+    _infoMap.addModelInfo(message);
+  }
+
+  private DataSchema parseSchema(File schemaFile) throws FileNotFoundException
+  {
+    DataSchemaResolver resolver = MultiFormatDataSchemaResolver.withBuiltinFormats(_resolverPath);
+    PdlSchemaParser parser = new PdlSchemaParser(resolver);
+    parser.parse(new FileInputStream(schemaFile));
+    if (parser.hasError())
+    {
+      throw new RuntimeException(parser.errorMessage() + " Error while parsing file: " + schemaFile.toString());
+    }
+
+    List<DataSchema> topLevelDataSchemas = parser.topLevelDataSchemas();
+    if (topLevelDataSchemas.size() != 1)
+    {
+      throw new RuntimeException("Could not parse schema : " + schemaFile.getAbsolutePath() + " The size of top level schemas is not 1.");
+    }
+    DataSchema topLevelDataSchema = topLevelDataSchemas.get(0);
+    if (!(topLevelDataSchema instanceof NamedDataSchema))
+    {
+      throw new RuntimeException("Invalid schema : " + schemaFile.getAbsolutePath() + ", the schema is not a named schema.");
+    }
+    return topLevelDataSchema;
+  }
+
+  private static String listCompatLevelOptions()
+  {
+    final StringBuilder options = new StringBuilder("<");
+    for (CompatibilityLevel compatLevel : CompatibilityLevel.values())
+    {
+      options.append(compatLevel.name().toLowerCase()).append("|");
+    }
+    options.replace(options.length() - 1, options.length(), ">");
+
+    return options.toString();
+  }
+
+  private static String listCompatModeOptions()
+  {
+    final StringBuilder options = new StringBuilder("<");
```

Comment:
We can try to use StringJoiner to replace StringBuilder, that will make this code a bit cleaner because we don't have to add delimiter every time when a element is appended.

You can have it this way
```
StringJoiner joiner = new StringJoiner("|", "<", ">");
Stream.of(CompatibilityOptions.Mode.values()).forEach(StringJoiner::add);
return joiner.toString();
```

Reply:
Updated based on the suggestion